Feature/webserver multi language

Baklap4 opened the pull request:
This merge request makes it possible to get the webserver frontend translated.

# Changes
* Switched from gettext to use boost::locale everywhere
* Added few test strings within the webserver
* Added api endpoint: `/api/language`
    * This endpoint serves json
    * The json can be used to translate everything within the client

# TODO
- [x] Switch from gettext to boost::locale (less code and threadsafe)
- [x] Add api endpoint `/api/language`
    - [x] API endpoint should serve json
- [x] Tested the singing part of the application regarding the new way to translate on Ubuntu 16.04
- [x] Add all default strings of the webfrontend
- [x] Retrieve translated strings within the webfrontend (by using jquery/javascript)
- [x] Optimize / clean up
    - [x] Cleaned up `index.html`
    - [x] Sorted every file out to their respecitivy folder: javascript in js, css in css and images in images.
    - [x] Remove gettext from mxe dependency list
- [x] Add endpoint to the api spec. Fixed (https://github.com/performous/performous/wiki/Webserver-API/731147678f20f76949846cef3f7582feeb7a21e7)
- [ ] Available Translations:
    - [ ] Asturian
    - [ ] Danish
    - [x] Spanish
    - [ ] Farsi
    - [ ] French
    - [ ] Hungarian
    - [ ] Italian
    - [ ] Finnish
    - [ ] Polish
    - [ ] Portuguese
    - [ ] Chinese
    - [ ] Swedish 
    - [x] Japanese
    - [x] Dutch 100%
    - [x] German 100%

# Involved issues
* Closes #188

TobiX commented:
You should probably also remove the Gettext detection from `game/CMakeLists.txt` to get rid of the library dependency.

And since you are already basically rewriting i18n.h, could you make ` _` a function instead of a macro?

Another thing: `using namespace` is bad C++ style, better avoid it.

(Oh, and of course remove your first try with pogettext from the commit history)

Baklap4 commented:
Just finished getting everything to work within the browser.

If anyone knows how to optimize this piece of javascript that'd be lovely : https://github.com/performous/performous/blob/bd512bdd17db6b156c6cbffda6484605f156c125/data/themes/default/www/js/site.js#L274-L281

I've picked up on @TobiX his sayings:
- Removed GetText detection files
- Removed `using namespace`

I still have a question regarding converting a macro to a function, help would be appreciated :)

Lord-Kamina commented:
From a cursory look, I'd say it's mostly fine; perhaps the loops could be
changed to use a better syntax but to be honest I'm not sure that's
actually more efficient.

On June 15, 2017 at 17:40:25, Arjan Spieard (notifications@github.com)
wrote:

> Just finished getting everything to work within the browser.
>
> If anyone knows how to optimize this piece of javascript that'd be lovely
> :
> https://github.com/performous/performous/blob/bd512bdd17db6b156c6cbffda6484605f156c125/data/themes/default/www/js/site.js#L274-L281
>
> I've picked up on @TobiX <https://github.com/tobix> his sayings:
>
>    - Removed GetText detection files
>    - Removed using namespace
>
> I still have a question regarding converting a macro to a function, help
> would be appreciated :)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/performous/performous/pull/236#issuecomment-308873414>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGJfLNWYDp2f9AKL2F5U7inM8uENZ1C3ks5sEaTJgaJpZM4N7ca6>
> .
>

Baklap4 commented:
@Lord-Kamina I'd say change all the loops to use the c++11 format?
For read/write actions:
```
for (auto& kv : myMap) {
    std::cout << kv.first << " has value " << kv.second << std::endl;
}
```

For readonly actions: 
```
for (auto const &kv : myMap) {
    std::cout << kv.first << " has value " << kv.second << std::endl;
}
```

same goes for vectors i think?

nieknooijens commented:
I'm more fond of actually typing the type name:

for (std::map const &kv : myMap)


Mainly because "auto" messes up qt-creators c++ autocomplete


2017-06-16 8:57 GMT+02:00 Arjan Spieard <notifications@github.com>:

> @Lord-Kamina <https://github.com/lord-kamina> I'd say change all the
> loops to use the c++11 format?
> For read/write actions:
>
> for (auto& kv : myMap) {
>     std::cout << kv.first << " has value " << kv.second << std::endl;
> }
>
> For readonly actions:
>
> for (auto const &kv : myMap) {
>     std::cout << kv.first << " has value " << kv.second << std::endl;
> }
>
> same goes for vectors i think?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/performous/performous/pull/236#issuecomment-308949408>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACqjWwQII358DyjXOxVYfcK_L8YHLunYks5sEidsgaJpZM4N7ca6>
> .
>



-- 
Met vriendelijke groet
Niek Nooijens

Baklap4 commented:
@nieknooijens personlly i'm quite fond of the `auto` keyword. But this might be because of the `var` keyword which we use a lot in c# (same purpose)

The whole reason for the `auto` keyword is readability and keeping the code 'clean'.

The issue with qt-creator i didn't know about. A quick google search brought me to this: https://stackoverflow.com/questions/14174409/no-autocompletion-support-in-qtcreator-with-auto-variable

nieknooijens commented:
@Baklap4 I actually think it's more readable if it has the right typename in there.
I hate stuff like javascript or python where everything is a var! the problem here is that as a programmer you have to search back to know what that variable was, and what it can be used for. I hate it in javascript where I do "var x + var y" and I get the error message "var x is a string, you can't add an int to it" how am I supposed to know? but if it has "string x + int y" then I immediately know what's going on.

the whole purpose of c++11 loop is to substitute:
```
for(int i=0; i<myMap.size(); i++) {
    std::map currentMap = myMap[i];
    currentMap.someOperation()
}

to

for (std::map currentMap : myMap) {
     currentMap.someOperation()
}

which is way easier to read, but substituting std::map for "auto" makes it more confusing again....

TobiX commented:
Please don't confuse dynamic typing (JavaScript) with type inference (C++ `auto`) - Idiomatic, modern C++ uses `auto` for loop variables since the type is cumbersome to write (`std::map<std::string, std::string>::iterator`) and quite clear for a developer familiar with modern C++ that it's an iterator over the type on the right.

nieknooijens commented:
@TobiX you don't need to use the ::iterator, then it implies you're having a list of iterators.

Baklap4 commented:
Soo @TobiX @nieknooijens what shall we use?

TobiX commented:
@nieknooijens What? Either you write `auto` or the full type... Can you give me an example for your claim? You could star with this example: https://godbolt.org/g/U3o12V

Baklap4 commented:
Also fyi the following javascript is correct and valid:
```
var x = "some string";
var y = 2;
var z = x+y;
console.log(z);
```
This will output: "some string2"

Baklap4 commented:
To end the discussion. I've used the `auto` keywords and changed the variables to readable names. I've also made them readonly by reference if i'm correct, which should be a bit faster than copying the whole object.

Baklap4 commented:
I'll rebase tonight didn't have time for it yet

Baklap4 commented:
@TobiX @nieknooijens  if no further objections i'll merge this with master in a day

nieknooijens commented:
@Baklap4 I scanned through the code and it looks like a great improvement to me, but to be able to accept this PR I first need to compile and test it myself which I will do tomorrow-evening. then I'll also add Japanese translation.

TobiX commented:
@Baklap4 Did you see my review regarding proper use of translation tools?

Baklap4 commented:
@TobiX Sadly i haven't :disappointed:

TobiX commented on game/webserver.cc:
```diff
@@ -203,6 +218,46 @@ http_server::response WebServer::POSTresponse(const http_server::request &reques
+		"Search for songs",
+		"Yes",
+		"No"
+	};
```
Let's try this again:

I'd suggest wrapping those in gettext_noop "functions", which would be defined as

    #define gettext_noop(x) x

(for example in i18n.h) and then setting up poedit as described in `lang/README`, adding gettext_noop in the supported keywords, so updating the performous.pot works automatically.

Baklap4 replied:
One question how would one get this specific set of 'keys' then? since i need them within `webserver.cc` or else the webserver doesn't get to send the json dictionairy (with these keys)

Currently we're putting everything in a `map<key (string), translated_value(string)>` and convert this map to a jsontree which then is used clientsided.

TobiX replied:
That's the whole point of a `gettext_noop` - It doesn't change the value, but serves as a marker for translation tools that this is a translatable string.

Baklap4 replied:
@TobiX you mean like this: https://github.com/performous/performous/pull/236/commits/e67a2fc4872c32cfbde4c7ae9844eb5736b16c2b ?

nieknooijens commented:
So I checked compilation and translation support by using `locale -a` to show my installed languages and then use `export LANG=<language from locale -a here>` to switch languages.
![schermafdruk van 2017-06-21 18-46-36](https://user-images.githubusercontent.com/2794331/27397672-68f57fd0-56b7-11e7-9549-19bc3e3d90bc.png)
dutch seems to work fine, so I fired up poedit opened Japanese, hit sync, and then fill in every empty space. result is not quite what I was expecting:
![schermafdruk van 2017-06-21 19-23-23](https://user-images.githubusercontent.com/2794331/27397728-97c3af80-56b7-11e7-87d2-2d1dd1176983.png)
so, for some reason not all strings are added to the Japanese .po file, which is weird. am I missing some undocumented config option?

Baklap4 commented:
@nieknooijens in a few i'll add a commit with suggestion from @TobiX so we won't miss strings
I am aware of one thing for example placeholders in html won't get the translated text since i didn't know how to fill in these placeholder values. I'll look into it

Baklap4 commented:
I'd like to know if this is mergeable: @nieknooijens @TobiX @Lord-Kamina 
Like all translations, it takes time to fill them all

Lord-Kamina commented:
Spanish, at least, is ok, and performous should build and launch ok on Mac. Any possible fixes I could do to my translation later on would be to address specific issues.

I still think we should have a configuration for language, though, as opposed to inferring it from the OS locale.

Baklap4 commented:
@Lord-Kamina Should it be part of this branch or as a new one? IMHO it'd be wise to create a different issue for this, since this is already a big commit to be done. This way we won't diverge much from master, and the possibility to get merge conflicts would decrease.

Lord-Kamina commented:
@Baklap4 Probably a new one.
P.S. just found a missing string in spanish. Under "Webserver Access" in the configuration screen, the different options are not translated.

Baklap4 commented:
@Lord-Kamina Push them when you have time :) I'll create an issue for the config option

nieknooijens commented on CMakeLists.txt:
```diff
@@ -6,7 +6,7 @@ set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake/Modules/")
 git_describe(PROJECT_VERSION --tags)
 if(NOT PROJECT_VERSION)
-	set(PROJECT_VERSION "1.1+")
+	set(PROJECT_VERSION "1.1.3")
```
this should actually be done in a release branch

nieknooijens commented on osx-utils/resources/Info.plist:
```diff
@@ -19,11 +19,11 @@
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>1.1</string>
+	<string>1.1.3</string>
```
again, release branch plz.

Baklap4 commented:
Anymore objections? Else i'll just hit the merge button... since it's already has 1 approval

TobiX commented:
Yes, the stray version bump that @nieknooijens commented, is still there. webserver.cc still contains one "untranslatable" string, which came about by imporoper use of translation tools.

 Also, IMHO, this should be split into multiple changes, currently this branch contains ~3 mostly not related changes which just came about by the initial idea, but that is just my suggestion for "cleaner" history.

Baklap4 commented:
@TobiX I'll split it in multiple pull requests, get the stray untranslateable string away, and remove the version bump.

I'll create 3 pull requests for the following:
1) Removing gettext
2) Updating webserver.cc + webfront
3) Translations

Sounds like plan?

Baklap4 commented:
Closing this since this in favor of the splitted pull requests
1. Replace GettText with Boost::Locale #262 
2. Make webserver localize ready #263 
3. Added translations #264
